File: ExpenseTracker.Api/Models/Invoice.cs
namespace ExpenseTracker.Api.Models;

public enum InvoiceStatus
{
    Draft,
    Sent,
    Paid,
    Overdue,
    Cancelled
}

public class Invoice
{
    public int Id { get; set; }
    public int UserId { get; set; }
    public int? CustomerId { get; set; }
    public string InvoiceNumber { get; set; } = string.Empty;
    public DateTime IssueDate { get; set; }
    public DateTime DueDate { get; set; }
    public InvoiceStatus Status { get; set; } = InvoiceStatus.Draft;
    public string CurrencyCode { get; set; } = "USD";
    public decimal TaxRate { get; set; }
    public string? Notes { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    public decimal Subtotal => Items.Sum(i => i.Amount);
    public decimal Tax => Math.Round(Subtotal * TaxRate / 100m, 2);
    public decimal Total => Subtotal + Tax;
    public decimal PaidAmount => Payments.Sum(p => p.Amount);
    public decimal Balance => Math.Max(Total - PaidAmount, 0);

    public User? User { get; set; }
    public Customer? Customer { get; set; }
    public ICollection<InvoiceItem> Items { get; set; } = new List<InvoiceItem>();
    public ICollection<InvoicePayment> Payments { get; set; } = new List<InvoicePayment>();
}

File: ExpenseTracker.Web/src/app/core/models/invoice.ts
export type InvoiceStatus = 'Draft' | 'Sent' | 'Paid' | 'Overdue' | 'Cancelled';
export type PaymentMethod = 'Cash' | 'Bank' | 'Card' | 'Other';

export interface Invoice {
  id: number;
  invoiceNumber: string;
  customerId: number | null;
  customerName: string | null;
  issueDate: string;
  dueDate: string;
  status: InvoiceStatus;
  currencyCode: string;
  taxRate: number;
  subtotal: number;
  tax: number;
  total: number;
  paidAmount: number;
  balance: number;
}

export interface InvoiceItem {
  id: number;
  description: string;
  quantity: number;
  unitPrice: number;
  amount: number;
}

export interface InvoicePayment {
  id: number;
  amount: number;
  paymentDate: string;
  method: PaymentMethod;
  reference: string | null;
}

export interface InvoiceDetail extends Invoice {
  notes: string | null;
  items: InvoiceItem[];
  payments: InvoicePayment[];
}

export interface InvoiceItemRequest {
  description: string;
  quantity: number;
  unitPrice: number;
}

export interface InvoiceRequest {
  customerId?: number | null;
  issueDate: string;
  dueDate: string;
  status: 'Draft' | 'Sent';
  taxRate: number;
  currencyCode: string;
  notes?: string | null;
  items: InvoiceItemRequest[];
}

export interface InvoiceQuery {
  status?: string;
  customerId?: number;
}

export interface PaymentRequest {
  amount: number;
  paymentDate: string;
  method: PaymentMethod;
  reference?: string | null;
}
